### PR TITLE
Milestone 2: User interface

### PR DESCRIPTION
**Pull request by phylischumba**
## In this milestone we:
- Create a new branch; user_interface
- Create two players; player_one and player_two
- Prompt user to choose  between "x" and "o"
- Draw a board to be used in the game
- Prompt players to choose positions on board and notify them of their choice
- Display board with corresponding value according to player-choice
- Display different win messages

### CODE REVIEW

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great start @team :clap: :clap: However, some requirements need to be implemented in order for us to approve the milestone. Check :arrow_down: 

##### Behaviour

* [x] In this step not everything must work, the point is to think about everything listed below. You might do it only with inline comments. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone2/logic.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
* [x] Basic game flow is prepared: ([screenshot hint](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone2/logic.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
  * [x] `[REQUIRED]` game informs player if selected move is invalid
  * [x] `[REQUIRED]` game informs player if selected move is a winning move
  * [x] `[REQUIRED]` game informs player if selected move is a draw move
  * [x] `[REQUIRED]` game repeats all actions for next player's move

##### Implementation details

- Your code isn't DRY enough:
  - The logic for displaying the board could be joined in a single method (e.g board_display(board) )
  - The game should interact with users continuously until it reaches a stopping condition. Below, we explain how you should implement the game loop.

* [x] A loop is needed in this game (recommended is `while` but anything else will do) ([OdinProject hint](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone2/loop.png))
* [x] Inside loop, All actions should be repeated for each player separately in each single iteration.
* [x] A loop needs to be terminated at some point (after a winning move or draw to be more specific). It can be done with variable:

```ruby
game_on = true

# initial instructions for player
current_player = player1
while game_on
  # loop for each move
  # display board
  # ask user for valid input
  if winner # and/or draw (the exact condition in this milestone is not important)
    game_on = false
else
  # switch current_player  
end
end

# You can display the game result after the loop
```

### That's it :relieved: When finished, kindly resubmit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

File: bin/main.rb [hunk 1]
```diff
@@ -1,3 +1,78 @@
...
+
+count = 0
+# game repeats all actions for player's moves
+while count <= 1
+  %w[one two].each do |i|
+    player = i
+
+    loop do
+      puts("\nPlayer #{i} It's your turn\n(choose between 1-9):")
+      choice = STDIN.gets.chomp.to_i
```
Comment (maelfosso) [inline bin/main.rb]: STDIN it's not needed. You could directly use `gets.chomp`

File: bin/main.rb [hunk 2]
```diff
@@ -1,3 +1,78 @@
...
+      # validate player move and choice
+      if game.player_choice?(choice)
+        if !game.pos_taken?(choice)
+          game.board_update(choice, player)
+          game.display_board
```
Comment (maelfosso) [inline bin/main.rb:60]: I don't know but as you could see in the screenshot uploaded above, there is no board displayed

Comment (maelfosso) [reply bin/main.rb:60]: After that step, you should say whether or not the game is in a win/draw state

File: bin/main.rb [hunk 3]
```diff
@@ -1,3 +1,78 @@
...
+  end
+end
+
+game = Instructions.new
+game.display_board
+
+count = 0
+# game repeats all actions for player's moves
+while count <= 1
+  %w[one two].each do |i|
```
Comment (maelfosso) [inline bin/main.rb:52]: I suggest you implement the method propose by @stratospherique

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello coders,
Thank you for the improvements,
However, it's not sufficient to move forward.

Please, keep in mind that the goal of this milestone is to set all the different parts of your project, I mean the backbone.
According to what you did, It missed a few things. Here is the execution of your app
Screenshot from 2020-04-23 14-53-43

[ ] After each turn, you didn't display the board
[ ] In your loop, I don't see anywhere the won or draw state game. After updating, the game you can just pick randomly 0/1 to decide if it's a win or not, just to show that you know the sequence of different steps
[ ] 

Also, please, read the comments below

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

File: bin/main.rb (hunk 1 above)
Comment (imahnama) [reply bin/main.rb]: Screenshot 2020-04-24 at 11 45 23
Hi @maelfosso Kindly confirm as seen below  in both my partner and I's screen the board displays after every move

**Comment by shshamim63** [comment]
## Status :herb: Approved
___
Hi @imahnama and @Phylis05 :wave: ,
The project contains all the messages to guide a user with a loop, though the message(related to draw) and loop contain some edge cases that should be avoided. But based on the main focus of this milestone are,
- contains all possible message to guide a user
- the developer is aware that it contains some repetitive tasks

I can't see any issue to block it. I just got a precaution to make. Don't think about generating a loop based on ```while variable < constant``` in your next milestone. Cause that is a clear indication that it is gonna violate the SOLID principles of OOP. If you need more info don't hesitate to knock me on slack.
___
##### This :+1: is from me, you can go ahead.
___
## :clubs: Happy Coding :clubs: 
### Shakhawat Hossain

**Pull request merged**